sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/MessageUtils.java

```diff
@@ -369,4 +380,189 @@ public static <T> T describedToOrigin(DescribedType describedType) {
+        } else {
+            throw new IllegalArgumentException("This API supports the addition of only SQLFilters and CorrelationFilters.");
+        }
+
+        RuleAction action = options.getAction();
+        if (action == null) {
+            descriptionMap.put(ManagementConstants.SQL_RULE_ACTION, null);
+        } else if (action instanceof SqlRuleAction) {
+            HashMap<String, Object> sqlActionMap = new HashMap<>();
+            sqlActionMap.put(ManagementConstants.EXPRESSION, ((SqlRuleAction) action).getSqlExpression());
+            descriptionMap.put(ManagementConstants.SQL_RULE_ACTION, sqlActionMap);
+        } else {
+            throw new IllegalArgumentException("This API supports the addition of only filters with SqlRuleActions.");
+        }
+
+        descriptionMap.put(ManagementConstants.RULE_NAME, name);
+
+        return descriptionMap;
+    }
+
+    /**
+     * Get message status from application properties.
+     *
+     * @param properties The application properties from message.
+     * @return Status code.
+     */
+    public static int getMessageStatus(ApplicationProperties properties) {
+        int statusCode = ManagementConstants.UNDEFINED_STATUS_CODE;
+        if (properties.getValue() == null) {
+            return statusCode;
+        }
+
+        Object codeObject = properties.getValue().get(ManagementConstants.STATUS_CODE);
+        if (codeObject == null) {
+            codeObject = properties.getValue().get(ManagementConstants.LEGACY_STATUS_CODE);
+        }
+        if (codeObject != null) {
+            statusCode = (int) codeObject;
+        }
+
+        return statusCode;
+    }
+
+    /**
+     * Get {@link RuleProperties} from {@link DescribedType}.
+     *
+     * @param ruleDescribedType A {@link DescribedType} with rule information.
+     * @return A {@link RuleProperties} contains name, {@link RuleAction} and {@link RuleFilter}.
+     */
+    @SuppressWarnings("unchecked")
+    public static RuleProperties decodeRuleDescribedType(DescribedType ruleDescribedType) {
+        if (ruleDescribedType == null) {
+            return null;
+        }
+        if (!(ruleDescribedType.getDescriptor()).equals(ServiceBusConstants.RULE_DESCRIPTION_NAME)) {
+            return null;
+        }
+
+        RuleProperties ruleProperties = new RuleProperties();
+        if (ruleDescribedType.getDescribed() instanceof ArrayList) {
+            ArrayList<Object> describedRule = (ArrayList<Object>) ruleDescribedType.getDescribed();
+            int count = describedRule.size();
+            if (count-- > 0) {
+                ruleProperties.setFilter(decodeFilter((DescribedType) describedRule.get(0)));
+            }
+
+            if (count-- > 0) {
+                ruleProperties.setAction(decodeRuleAction((DescribedType) describedRule.get(1)));
+            }
+
+            if (count > 0) {
+                ruleProperties.setName((String) describedRule.get(2));
+            }
```

Comment:
Iterator might be a better pattern in this code (than count-- and still index by hardcoded number).

Reply:
I agree with weidong, this is a bit hard to follow.